src/ai/ai.cpp

```diff
@@ -12,6 +12,231 @@
 
 namespace ai {
 
+namespace alliance_factors {
+/*
+	Returns `true` if `other` has one of `n`'s cores
+*/
+bool has_our_core(sys::state& state, dcon::nation_id n, dcon::nation_id other) {
+	for(auto prov : state.world.nation_get_province_ownership(other)) {
+		auto fat = fatten(state.world, prov).get_province();
+		for(auto c : fat.get_core_as_province()) {
+			auto core = fatten(state.world, c);
+			if(core.get_identity().get_nation_from_identity_holder() == n)
+				return true;
+		}
+	}
+	return false;
+}
+
+bool too_many_gp_allies(sys::state& state, dcon::nation_id target, dcon::nation_id from) {
+	uint32_t upper_gp_ally_limit = (state.military_definitions.great_wars_enabled) ? uint32_t(2) : uint32_t(1);
+	uint32_t gp_ally_count = uint32_t(0);
+
+	std::vector<dcon::nation_id> allies;
```

Comment:
I would rather not have what feels like an unnecessary vector doing allocations here. Just looping over the diplomatic relationships and checking for gp allies directly would be more efficient. If you want to factor out duplicate code from that process, you can write an iterator that iterates over just the allies of a nation, allowing you to write `for(auto a : alliesof(state, from)) { ... }` which then allows you to even break out early if you hit the upper_gp_ally_limit